### PR TITLE
New Encounters

### PR DESCRIPTION
**Pull request by wersal454**
## What type of PR is this?
1. [ ] Bug
2. [ ] Change
3. [x] Enhancement
4. [ ] Miscellaneous

### What have you changed and why?
Information:
Added new encounters:
Civilian convoy fleeing the city 
Slingload heli carrying some sweet sweet ammo boxes/containers/pods or vehicles
Frontline skirmish between Occ and Inv
Skirmish between Rivals and Occ

Other changes:
adjusted skill for rivals, because it turns out they can't hit shit
Made it so civ vehicles will now spawn driver from template (A.I. the "unitMan") and not some rando vanilla guy

### Please verify the following.

1. [x] Have you loaded the mission in LAN host?
2. [ ] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?

1. [ ] No
2. [x] Yes (Please provide further detail below.)

### How can the changes be tested?
Steps:
sling load heli encounter
[[], "SCRT_fnc_encounter_HeliSlingloadCargo"] call A3A_fnc_scheduler;

frontline skirmish
[[], "SCRT_fnc_encounter_frontlineSkirmish"] call A3A_fnc_scheduler;

civilian convoy
[[], "SCRT_fnc_encounter_fleeingCivConvoy"] call A3A_fnc_scheduler;

Rivals vs Occ skirmish
(In order to spawn that one you need rivals to be active) 
[[], "SCRT_fnc_rivals_encounter_OccVsRivalsSkirmish"] call A3A_fnc_scheduler;

If you don't have a save with active rivals use this commands
[] remoteExecCall ["SCRT_fnc_rivals_activate", 2];
then
spam this
[] spawn SCRT_fnc_rivals_activityUpdateLoop;
********************************************************
Notes:
Convoy is basicly ready, some adjustments needs to be made, like end of encounter condition
Same for sling load heli, minor bugs may pop up , + maybe add some functionality to it, like reinforce garrison or something
Skirmishes need some work, like make it spawn units properly and conditions for ending + balance adjustments (I left notes in both functions)

### CODE REVIEW

**Comment by Silence-112** [comment]
Change base branch to unstable please

File: A3A/addons/core/Templates/Templates/Vanilla/Vanilla_AI_AAF.sqf [hunk 1]
```diff
@@ -214,7 +214,6 @@ if (_hasCSLA) then {
 ["vehiclesLightArmed", _armedVehicles] call _fnc_saveToTemplate;
 ["vehiclesLightTanks",  _lightTanks] call _fnc_saveToTemplate;
 ["vehiclesAirborne", _airborneVehicles] call _fnc_saveToTemplate;
-["vehiclesMilitiaAPCs", _militiaAPCs] call _fnc_saveToTemplate;
```
Comment (Silence-112) [inline A3A/addons/core/Templates/Templates/Vanilla/Vanilla_AI_AAF.sqf:217]: any particular reason this was removed?

Comment (wersal454) [reply A3A/addons/core/Templates/Templates/Vanilla/Vanilla_AI_AAF.sqf:217]: saved twice

File: A3A/addons/scrt/Stringtable.xml [hunk 2]
```diff
@@ -1494,11 +1498,15 @@
                 <Chinesesimp>武装直升机</Chinesesimp>
                 <Korean>건쉽</Korean>
             </Key>
-            <Key ID="STR_marker_arty_target">
```
Comment (Silence-112) [inline A3A/addons/scrt/Stringtable.xml]: Why has this been removed/changed?

File: A3A/addons/scrt/Encounter/fn_encounter_HeliSlingloadCargo.sqf [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+};
+private _spawnPosition = selectRandom _potentialspawnPosition;
+if (_spawnPosition == _outpost) then {
+	while {true} do {
+		_potentialspawnPosition = (outposts + milbases + airportsX + resourcesX + factories) select {
+    	sidesX getVariable [_x, sideUnknown] != teamPlayer && sidesX getVariable [_x, sideUnknown] == _side  &&  {(getMarkerPos _x) distance2D _player < distanceSPWN}};
+	};
+	_spawnPosition = selectRandom _potentialspawnPosition;
+	if (_spawnPosition != _outpost) exitwith{};
+};
```
Comment (Silence-112) [inline A3A/addons/scrt/Encounter/fn_encounter_HeliSlingloadCargo.sqf]: What does this do? It seems unsafe

File: A3A/addons/scrt/Encounter/fn_encounter_HeliSlingloadCargo.sqf [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+
+if (_HeliClass == "") exitWith {
+    Error("No Helicopters, problem with template, aborting Heli slingload cargo Event.");
+    isEventInProgress = false;
+    publicVariableServer "isEventInProgress";
+};
+
+if (_HeliClass isKindOf "Heli_Transport_04_base_F") then {
+	_HeliClass = "O_Heli_Transport_04_F";
+};
```
Comment (Silence-112) [inline A3A/addons/scrt/Encounter/fn_encounter_HeliSlingloadCargo.sqf:68]: Why is this needed?

File: A3A/addons/scrt/Encounter/fn_encounter_HeliSlingloadCargo.sqf [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+private _heliVehicleData = [_actualspawnPosition, 90, _HeliClass, _side] call A3A_fnc_spawnVehicle;
+
+private _heliVehicle = _heliVehicleData select 0;
+private _lootCratetest = [_ammoBoxType, _actualspawnPosition, 20, 5, true] call A3A_fnc_safeVehicleSpawn;
+if !(_heliVehicle canSlingLoad _lootCratetest) exitwith {
+    Error("Your heli can't do shit, aborting.");
+    isEventInProgress = false;
+    publicVariableServer "isEventInProgress";
+};
+deleteVehicle _lootCratetest;
```
Comment (Silence-112) [inline A3A/addons/scrt/Encounter/fn_encounter_HeliSlingloadCargo.sqf]:
```suggestion
private _lootCrateTest = [_ammoBoxType, _actualspawnPosition, 20, 5, true] call A3A_fnc_safeVehicleSpawn;
if !(_heliVehicle canSlingLoad _lootCrateTest) exitwith {
    Error("Your heli can't do shit, aborting.");
    isEventInProgress = false;
    publicVariableServer "isEventInProgress";
};
deleteVehicle _lootCrateTest;
```

Follow the camel case standard

File: A3A/addons/scrt/Stringtable.xml (hunk 2 above)
Comment (wersal454) [reply A3A/addons/scrt/Stringtable.xml]: Accidently

File: A3A/addons/scrt/Encounter/fn_encounter_HeliSlingloadCargo.sqf [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+private _lootcrateType = "";
+private _lootCrate = objNull;
+private _attempts = 15;
+if (_HeliClass == "O_Heli_Transport_04_F") then {
+	_HeliClass = "O_Heli_Transport_04_F";
+	_lootcrateType = selectRandom ((_faction get "vehiclesLightUnarmed") + (_faction get "vehiclesLightArmed") + (_faction get "vehiclesAirborne") + 
+	(_faction get "vehiclesAA") + (_faction get "vehiclesLightTanks") + (_faction get "vehiclesMilitiaAPCs") + (_faction get "vehiclesAPCs") + (_faction get "vehiclesIFVs") + 
+	["Land_Pod_Heli_Transport_04_covered_F"] + ["Land_Pod_Heli_Transport_04_bench_F"] + ["Land_Pod_Heli_Transport_04_medevac_F"] + ["Land_Pod_Heli_Transport_04_repair_F"] + ["Land_Pod_Heli_Transport_04_fuel_F"] + 
+	["Land_Pod_Heli_Transport_04_ammo_F"] + ["Land_Pod_Heli_Transport_04_box_F"]);
+	while {_attempts != 0 } do {
```
Comment (Silence-112) [inline A3A/addons/scrt/Encounter/fn_encounter_HeliSlingloadCargo.sqf]:
```suggestion
	while {_attempts != 0} do {
```

File: A3A/addons/scrt/Encounter/fn_encounter_HeliSlingloadCargo.sqf [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+			} forEach units _heliInfGroup;
+			{
+			private _index = _lootCrate getCargoIndex _x;
+				if (_index == -1) then {
+    			deleteVehicle _x;
+				};
+			}forEach units _heliInfGroup;
+		};
+		if (_lootcrateType != "Land_Pod_Heli_Transport_04_medevac_F" && _lootcrateType != "Land_Pod_Heli_Transport_04_repair_F" && _lootcrateType != "Land_Pod_Heli_Transport_04_fuel_F" &&
+		 	_lootcrateType != "Land_Pod_Heli_Transport_04_ammo_F" && _lootcrateType != "Land_Pod_Heli_Transport_04_box_F" && _lootcrateType != "Land_Pod_Heli_Transport_04_covered_F" && _lootcrateType != "Land_Pod_Heli_Transport_04_bench_F") then {
```
Comment (Silence-112) [inline A3A/addons/scrt/Encounter/fn_encounter_HeliSlingloadCargo.sqf]: Put the classnames in an array, then do a simplified check:

```
_lootCrateTypes = ["...", "..."]

if (_lootCrateType in _lootCrateTypes) then { // ...
```

File: A3A/addons/scrt/Encounter/fn_encounter_HeliSlingloadCargo.sqf [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+			_lootCrate = [_lootcrateType, _actualspawnPosition, 40, 5, true] call A3A_fnc_safeVehicleSpawn;
+			[_lootCrate, _side] call A3A_fnc_AIVEHinit;
+			[_lootCrate] spawn A3A_fnc_fillLootCrate;
+			[_lootCrate] call A3A_Logistics_fnc_addLoadAction;
+		} else {
+			_lootCrate = [_lootcrateType, _actualspawnPosition, 40, 5, true] call A3A_fnc_safeVehicleSpawn;
+			[_lootCrate, _side] call A3A_fnc_AIVEHinit;
+			[_lootCrate] call A3A_Logistics_fnc_addLoadAction;
+		};
+		if (_lootcrateType != "B_Slingload_01_Ammo_F" && _lootcrateType != "B_Slingload_01_Medevac_F" && _lootcrateType != "B_Slingload_01_Repair_F" && _lootcrateType != "B_Slingload_01_Fuel_F" && _lootcrateType != "B_Slingload_01_Cargo_F") then {
```
Comment (Silence-112) [inline A3A/addons/scrt/Encounter/fn_encounter_HeliSlingloadCargo.sqf]: Do the same ```if (thing in listOfThings) then {``` here

**Review by Silence-112 — CHANGES_REQUESTED** [review]

File: A3A/addons/scrt/Encounter/fn_encounter_HeliSlingloadCargo.sqf (hunk 4 above)
Comment (wersal454) [reply A3A/addons/scrt/Encounter/fn_encounter_HeliSlingloadCargo.sqf:68]: CSAT helis that are in transport category, can't slingload because they already have pod, kinda, so I'm swapping them to actual transport heli

File: A3A/addons/scrt/Encounter/fn_encounter_HeliSlingloadCargo.sqf (hunk 3 above)
Comment (wersal454) [reply A3A/addons/scrt/Encounter/fn_encounter_HeliSlingloadCargo.sqf]: Checks for outposts in spawn distance for AI, though I need a recheck, update comment later

Comment (wersal454) [reply A3A/addons/scrt/Encounter/fn_encounter_HeliSlingloadCargo.sqf]: It gets another outpost that is != spawnoutpost, so heli has somewhere to fly

Comment (Silence-112) [reply A3A/addons/scrt/Encounter/fn_encounter_HeliSlingloadCargo.sqf]: Using a while {true} loop is unsafe as it has no exit condition (in this particular instance). Essentially, you've got code running every (frame?) even when it's no longer needed.

I think I see what you were going for with the exitWith, shouldn't that be inside the while loop?

File: A3A/addons/scrt/Encounter/fn_encounter_HeliSlingloadCargo.sqf (hunk 4 above)
Comment (Silence-112) [reply A3A/addons/scrt/Encounter/fn_encounter_HeliSlingloadCargo.sqf:68]: Fair enough

File: A3A/addons/scrt/Encounter/fn_encounter_HeliSlingloadCargo.sqf [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+private _lootCrate = objNull;
+private _attempts = 15;
+
+if (_HeliClass == "O_Heli_Transport_04_F") then {
+	_HeliClass = "O_Heli_Transport_04_F";
+	_lootcrateType = selectRandom ((_faction get "vehiclesLightUnarmed") + (_faction get "vehiclesLightArmed") + (_faction get "vehiclesAirborne") + 
+	(_faction get "vehiclesAA") + (_faction get "vehiclesLightTanks") + (_faction get "vehiclesMilitiaAPCs") + (_faction get "vehiclesAPCs") + (_faction get "vehiclesIFVs") + _csatPods);
+	while {_attempts != 0} do {
+		_lootcrateType = selectRandom ((_faction get "vehiclesLightUnarmed") + (_faction get "vehiclesLightArmed") + (_faction get "vehiclesAirborne") + (_faction get "vehiclesAA") + (_faction get "vehiclesLightTanks") + 
+		(_faction get "vehiclesMilitiaAPCs") + (_faction get "vehiclesAPCs") + (_faction get "vehiclesIFVs") + + _csatPods);
```
Comment (Silence-112) [inline A3A/addons/scrt/Encounter/fn_encounter_HeliSlingloadCargo.sqf]:
```suggestion
		(_faction get "vehiclesMilitiaAPCs") + (_faction get "vehiclesAPCs") + (_faction get "vehiclesIFVs") + _csatPods);
```

File: A3A/addons/scrt/Encounter/fn_encounter_fleeingCivConvoy.sqf [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+    Info("No cities in proximity, aborting civ convoy Event.");
+    isEventInProgress = false;
+    publicVariableServer "isEventInProgress";
+};
+
+private _anothercity = selectRandom citiesX;
+while {true} do {
+	if (_anothercity != _city) exitwith{};
+	private _anothercity = selectRandom citiesX;
+};
```
Comment (Silence-112) [inline A3A/addons/scrt/Encounter/fn_encounter_fleeingCivConvoy.sqf]:
```suggestion
private _anotherCity = selectRandom (citiesX select {_x != _city})
```

Can't this achieve the same thing?
Make sure to test it as i'm not entirely sure, but these rogue while loops are terrible to use

**Review by Silence-112 — COMMENTED** [review]
Slingload is fine as far as I can tell

I've tested all of them and they're generally fine, my only issues are:

Civ fleeing convoy seems to tank performance, might have something to do with the while loop;

Rivals skirmish can be activated without rivals active. Might not be an issue if the choosing function takes rivals activated into account;

Frontline skirmish is kind of lackluster, and they spawn way too close to each other. That might be due to the error that me and wersal discussed in discord

```sqf
14:26:07 Error in expression <hposition = [_FrontlineOutpostPosition, distanceSPWN0, distanceSPWN*1.1, 10, 0, >
14:26:07   Error position: <distanceSPWN0, distanceSPWN*1.1, 10, 0, >
14:26:07   Error Undefined variable in expression: distancespwn0
14:26:07 File x\A3A\addons\scrt\Encounter\fn_encounter_frontlineSkirmish.sqf..., line 66
14:26:07  ➥ Context:     [] L66 (x\A3A\addons\scrt\Encounter\fn_encounter_frontlineSkirmish.sqf)
```

Other than that, I think it's fine

**Review by Silence-112 — APPROVED** [review]
